h2o-py/h2o/model/model_base.py

```diff
@@ -234,16 +234,14 @@ def predict_contributions(self, test_data, output_format="Original", top_n=None,
         >>> # Compute SHAP and pick the top two highest and top two lowest
         >>> m.predict_contributions(fr, top_n=2, bottom_n=2)
         """
-        assert_is_type(output_format, None, Enum("Original", "Compact"))
-        if not isinstance(test_data, h2o.H2OFrame): raise ValueError("test_data must be an instance of H2OFrame")
-        j = H2OJob(h2o.api("POST /4/Predictions/models/%s/frames/%s" % (self.model_id, test_data.frame_id),
-                           data={"predict_contributions": True,
-                                 "predict_contributions_output_format": output_format,
-                                 "top_n": top_n,
-                                 "bottom_n": bottom_n,
-                                 "compare_abs": compare_abs}), "contributions")
-        j.poll()
-        return h2o.get_frame(j.dest_key)
+        if has_extension(self, 'PredictContributions'):
+            return self._predict_contributions(test_data, output_format, top_n, bottom_n, compare_abs)
+        warn_msg = "WARNING: This model doesn't support calculation of feature contributions."
+        if has_extension(self, 'StandardCoef'):
+            warn_msg += " When features are independent, you can use the coef() method to get coefficients"
+            warn_msg += " for non-standardized data or coef_norm() to get coefficients for standardized data."
+            warn_msg += " You can plot standardized coefficient magnitudes by calling std_coef_plot() on the model."
+        print(warn_msg)
```

Comment:
so, if the model doesn't support contributions, we prefer returning `None` and print a message in the console rather than emitting a proper error?
Meaning if client does:
```python
contributions = model.prediction_contributions(...)
len(contributions)
```
the user will see sth like
```
TypeError                                 Traceback (most recent call last)
<ipython-input-1-96b07938784c> in <module>
----> 1 len(None)

TypeError: object of type 'NoneType' has no len()
```
instead of sth like
```
H2OUnsupportedError                                 Traceback (most recent call last)
<ipython-input-1-96b07938784c> in <module>
----> 1 contributions = model.prediction_contributions(...)

H2OUnsupportedError: This model doesn't support calculation of feature contributions.
```

Reply:
I think we've been using the `print(warning)` + `return None` logic in various places, but I still have no idea why we went down that road...

Reply:
I have tried to be consistent with other parts of python API. 
- https://github.com/h2oai/h2o-3/blob/master/h2o-py/h2o/model/model_base.py#L479
- https://github.com/h2oai/h2o-3/blob/master/h2o-py/h2o/model/model_base.py#L523
- https://github.com/h2oai/h2o-3/blob/master/h2o-py/h2o/model/model_base.py#L565
...

If the method should throw an error, IMHO should do the same.

Reply:
@mn-mikke I agree with you. When I created the extensions logic, I wanted to completely remove the methods from `model_base.py`, this would have exposed the methods only to models supporting the feature...
But due to legacy logic, and to stay backwards compatible, we kept this `print(warning)` logic instead.

However, it looks like `predict_contributions` was already failing anyway for models not supporting contributions, so here, I'd completely remove the method from `model_base.py` and make the `predict_contributions` in the extension public... this would be much simpler imo.
But also, in this PR, it seems like you need to add an extended message for models that don't support contributions but do support standardized coefficients: this looks a bit strange to me, sth between the error message and the user tip, "you tried to do that, but you can't, so we'd suggest you to try this instead and we inform you before returning None so that you're likely to crash in the next statement..."

My 2 cents:
- let's remove `predict_contributions` from `model_base.py`.
- make `_predict_contributions` in the extension public.
- add a `predict_contributions` method to `StandardCoef` extension that just raises an error with the additional tip above.
- if there were algos that supported both contributions and standardized coef (there's none today as GLM doesn't support contributions), we would just need to ensure that the `Contributions` one is declared before the `StandardCoef` one.
This way, no method for algos that don't support contributions, except for GLM (StandardCoef) that raises a more detailed error.
wdyt?

Reply:
@sebhrusen I've modified the PR to throw `H2OTypeException` instead:
<img width="998" alt="Screen Shot 2022-08-29 at 20 07 17" src="https://user-images.githubusercontent.com/3674621/187269615-ca29516e-d6f8-4972-b0f7-ae2d7cc8bb32.png">